### PR TITLE
Slack: route subagent completion messages to originating thread

### LINKED ISSUE #64454: [Bug]: Slack extension: subagent completion messages bypass thread routing (missing subagent_delivery_target hook)
### Bug type

Behavior bug (incorrect output/state without crash)

### Beta release blocker

No

### Summary

## Summary

When the Slack channel plugin spawns a subagent from within a thread, the subagent's completion message is delivered to the main channel instead of the originating thread. Discord and Feishu handle this correctly; Slack is missing the hook.

**Version:** 2026.4.9

## Repro

1. In a Slack channel with the bot active, reply inside an existing thread with a question that triggers subagent delegation (e.g. "check the X API and tell me Y").
2. The router replies in-thread with a status message (e.g. "Checking now — one moment") — correct, lands in thread.
3. Router delegates to a subagent (`worker-deep`, `worker-fast`, etc.).
4. When the subagent completes, the result is posted to the **main channel**, not the thread.

## Expected

Subagent completion result delivered back into the originating thread, same as Discord and Feishu.

## Evidence

The thread ID is captured correctly in the subagent run record — this is not a plumbing issue. From `~/.openclaw/subagents/runs.json`:

```json
{
  "controllerSessionKey": "agent:router:slack:channel:<channel>:thread:<thread_ts>",
  "requesterSessionKey": "agent:router:slack:channel:<channel>:thread:<thread_ts>",
  "requesterOrigin": {
    "channel": "slack",
    "to": "channel:<channel>",
    "accountId": "default",
    "threadId": "<thread_ts>"
  },
  "expectsCompletionMessage": true
}
```

The `threadId` is stored and `expectsCompletionMessage: true` is set, but nothing reads the threadId back when the completion fires. Setting `channels.slack.replyToMode: "all"` does not fix this — the subagent delivery path bypasses `replyToMode` entirely.

## Root cause

The Slack extension does not register a `subagent_delivery_target` hook. Discord and Feishu both do:

- `dist/extensions/discord/index.js:30` — `api.on("subagent_delivery_target", …)` → `handleDiscordSubagentDeliveryTarget`
- `dist/extensions/feishu/index.js:66` — same pattern
- `dist/extensions/slack/*.js` — **no `api.on` calls, no subagent hooks registered**

The hook runner at `dist/hook-runner-global-BR8jhpWH.js:477` (`runSubagentDeliveryTarget`) runs the `subagent_delivery_target` modifying hook; with no Slack handler, nothing overrides the default delivery target, which drops the thread context.

The Discord reference implementation at `dist/subagent-hooks-DG3cIH2I.js:74-99`:

```js
function handleDiscordSubagentDeliveryTarget(event) {
    if (!event.expectsCompletionMessage) return;
    if (normalizeOptionalLowercaseString(event.requesterOrigin?.channel) !== "discord") return;
    const requesterAccountId = event.requesterOrigin?.accountId?.trim();
    const requesterThreadId = event.requesterOrigin?.threadId != null && event.requesterOrigin.threadId !== ""
        ? normalizeOptionalStringifiedId(event.requesterOrigin.threadId) ?? ""
        : "";
    const bindings = listThreadBindingsBySessionKey({
        targetSessionKey: event.childSessionKey,
        …requesterAccountId ? { accountId: requesterAccountId } : {},
        targetKind: "subagent"
    });
    // … resolve binding, return { origin: { channel, accountId, to, threadId } }
}
```

Discord also registers a spawn-time binding (`handleDiscordSubagentSpawning` → `autoBindSpawnedDiscordSubagent`) so the completion-time lookup has something to find. Slack is missing both pieces.

## Suggested fix

1. Add `handleSlackSubagentSpawning` modeled on the Discord version, calling an equivalent `autoBindSpawnedSlackSubagent` that writes a thread binding at spawn time.
2. Add `handleSlackSubagentDeliveryTarget` modeled on `handleDiscordSubagentDeliveryTarget`, reading `requesterOrigin.threadId` and returning the Slack thread as the delivery target.
3. Register both hooks in the Slack plugin's API wiring (parallel to `extensions/discord/index.js:30`).

The Slack thread ID format (`1775841953.287659`) is directly usable as `thread_ts` in `chat.postMessage`, so the binding wants to set `thread_ts` on outbound `postMessage` calls via the existing `replyThreadTs` / `payload.replyToId` pathway in `dist/replies-Dv7Xeg_O.js:15`.

## Workarounds

- None at the config level. `replyToMode: "all"` helps unthreaded direct replies but does not reach the subagent completion path.
- Avoid thread-scoped subagent delegation until fixed.

### Steps to reproduce

1. Run OpenClaw 2026.4.9 with `channels.slack` configured (socket mode, `replyToMode: "all"`) and a multi-agent setup where   
  the router can delegate to `worker-deep` via subagent spawn.                                                                  
  2. From an allowlisted Slack user, post a message inside an existing Slack thread that prompts the router to delegate research
   to a subagent (e.g., "Can ShipHero set holds based on date in bulk?" inside a thread).                                       
  3. Observe the router's inline reply land in the thread correctly, followed by a "Checking… one moment" acknowledgement (also 
  in-thread).                                                                                                                   
  4. Wait for the subagent to complete (run label `worker-deep-shiphero-bulk-hold-research`, ~60s in the observed trace).       
  5. Observe the subagent's completion message posted to the main channel instead of the originating thread.        

### Expected behavior

Within the same session, the router's direct in-thread reply was delivered correctly into the thread (Slack message ts        
  `1775843604.449789`, thread root `1775841953.287659`). The subagent completion message was expected to use the same thread as 
  the delivery target.

### Actual behavior

  The subagent completion message was posted to the main channel, not the thread. Slack message ts `1775843680.428899` (no
  `thread_ts`), while the originating thread root is `1775841953.287659`. Timing correlates with the subagent run's
  `endedHookEmittedAt: 1775843680699` in `~/.openclaw/subagents/runs.json` (run id `c7d7d892-cb02-4ecc-a234-0e7cea5d6c8b`),
  matching the Slack post time within ~300ms. 

### OpenClaw version

2026.4.9

### Operating system

macOS 26.3.1 (build 25D2128, Darwin 25.3.0)

### Install method

npm global under Homebrew Node v25.5.0 (installed at /opt/homebrew/lib/node_modules/openclaw)

### Model

Subagent (worker-deep) that produced the misrouted message: anthropic/claude-opus-4-6. Router: anthropic/claude-sonnet-4-6.   The misrouting occurs in the delivery path after model inference and is independent of the model.    

### Provider / routing chain

openclaw -> Anthropic API (direct, api_key auth profile, no gateway/proxy) 

### Additional provider/model setup details

`auth.profiles.anthropic:default` set to `{ provider: "anthropic", mode: "api_key" }` in `~/.openclaw/openclaw.json`. No
  custom baseUrl, no intermediate routing. `channels.slack` uses socket mode with `replyToMode: "all"` set. Slack plugin enabled
   via `plugins.entries.slack.enabled = true`.

### Logs, screenshots, and evidence

```shell
Subagent run record from `~/.openclaw/subagents/runs.json` (fields trimmed, identifiers redacted):
                                                                                                    
  {
    "runId": "c7d7d892-cb02-4ecc-a234-0e7cea5d6c8b",                                                                            
    "childSessionKey": "agent:worker-deep:subagent:56c82ad3-7946-4c65-8eea-fddb7054476e",
    "controllerSessionKey": "agent:router:slack:channel:<channel>:thread:1775841953.287659",                                    
    "requesterSessionKey": "agent:router:slack:channel:<channel>:thread:1775841953.287659",                                     
    "requesterOrigin": {                                                                                                        
      "channel": "slack",                                                                                                       
      "to": "channel:<channel>",                                                                                                
      "accountId": "default",                                                                                                   
      "threadId": "1775841953.287659"
    },                                                                                                                          
    "label": "worker-deep-shiphero-bulk-hold-research",                                                                         
    "expectsCompletionMessage": true,                  
    "spawnMode": "run",                                                                                                         
    "startedAt": 1775843598430,                                                                                                 
    "endedAt": 1775843665031,  
    "outcome": { "status": "ok" },                                                                                              
    "endedReason": "subagent-complete",
    "endedHookEmittedAt": 1775843680699                                                                                         
  }                                                                                                                             
   
  Note `requesterOrigin.threadId` is captured at spawn time and `expectsCompletionMessage: true` is set, but the thread is      
  dropped at delivery.                                                                                                          
                                                                                                                                
  Code inspection of the installed package at `/opt/homebrew/lib/node_modules/openclaw/dist`:                                   
                                                                                                                                
  - `extensions/discord/index.js:30` — `api.on("subagent_delivery_target", ...) -> handleDiscordSubagentDeliveryTarget`
  - `extensions/feishu/index.js:66` — same pattern                                                                              
  - `extensions/slack/*.js` — no `api.on(...)` calls anywhere; no subagent hooks registered (verified by grep for `api.on` and
  `subagent` across the Slack extension directory)                                                                              
  - `subagent-hooks-DG3cIH2I.js:74-99` — Discord's `handleDiscordSubagentDeliveryTarget` reads `event.requesterOrigin.threadId` 
  and returns `{ origin: { channel, accountId, to, threadId } }` so completion delivery routes back into the thread. No Slack   
  equivalent exists.                                                                                                            
  - `hook-runner-global-BR8jhpWH.js:477` — `runSubagentDeliveryTarget` runs the `subagent_delivery_target` modifying hook; with
  no Slack handler, nothing overrides the default target, which drops the thread context.                                       
                                                                                                                                
  Slack message timestamps (same session):
  - `1775843604.449789` — router in-thread status ("Checking the ShipHero API now — one moment") — correctly threaded           
  - `1775843680.428899` — subagent completion result — posted to main channel, not thread                            
                                                                                                                                
  The config-level `channels.slack.replyToMode: "all"` setting does not reach this path; the subagent delivery pipeline resolves
   its own target via the `subagent_delivery_target` hook chain, bypassing `replyToMode`.
```

### Impact and severity

Affected: Slack channels where the OpenClaw router delegates to subagents from within an existing thread.
  Severity: Blocks thread workflow. The initiating user sees the router acknowledge in-thread ("Checking... one moment"), then
  the actual answer appears in the main channel, visually disconnected from the question.                                     
  Frequency: 1/1 observed attempts in this configuration.                                                                       
  Consequence: Subagent-produced answers are orphaned from the thread that triggered them. Readers of the thread see an
  unresolved "checking…" message; readers of the main channel see an answer with no visible question.    

### Additional information

 Same-session comparison: direct router replies in the thread were delivered correctly via the `replyThreadTs` /
  `payload.replyToId` pathway in `dist/replies-Dv7Xeg_O.js:15`. Only the subagent completion path is affected.   
                                                                                                              
  Suggested fix (grounded by the Discord reference implementation):                                                             
  1. Add a `handleSlackSubagentSpawning` that writes a thread binding at spawn time (parallel to                                
  `autoBindSpawnedDiscordSubagent`).                                                                                            
  2. Add a `handleSlackSubagentDeliveryTarget` that reads `event.requesterOrigin.threadId` and returns a Slack thread as the    
  delivery target (parallel to `handleDiscordSubagentDeliveryTarget` at `dist/subagent-hooks-DG3cIH2I.js:74-99`).           
  3. Register both hooks in the Slack plugin API wiring, parallel to `extensions/discord/index.js:30`.                          
                                                                                                      
  Regression window: NOT_ENOUGH_INFO. No prior OpenClaw version was tested against this workflow in the observed environment, so
   last-known-good and first-known-bad versions cannot be grounded.

### PR DESCRIPTION
**Pull request by jaredar1232**
## Summary

- Problem: Slack subagent completion messages bypass thread routing and land in the main channel instead of the originating Slack thread. Discord, Feishu, Telegram, and Matrix already handle this correctly; Slack was the outlier.
- Why it matters: Users expect subagent replies to continue in the same thread the request came from, and dropping replies into the main channel is noisy and breaks the conversational context.
- What changed: Added a Slack thread-binding manager plus the three subagent hooks (`subagent_spawning`, `subagent_ended`, `subagent_delivery_target`) following the same adapter-plus-hooks pattern used by the other channel plugins. The Slack channel plugin now exposes a `conversationBindings` block with Slack-specific lifecycle wiring, and the monitor provider instantiates a per-account manager when the feature is enabled.
- What did NOT change: Slack direct-message send paths, any non-subagent routing behavior, Discord/Telegram/Matrix/Feishu subagent routing, or the core subagent delivery logic in `src/agents/subagent-announce-delivery.ts`.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [x] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #64454
- Related #
- [x] This PR fixes a bug or regression

## Root Cause (if applicable)

- Root cause: Slack's channel plugin did not implement the conversation-binding adapter or the three subagent hook handlers (`subagent_spawning`, `subagent_ended`, `subagent_delivery_target`) that other channel plugins use to stamp and re-resolve thread origin metadata across the subagent lifecycle. The shared `resolveSubagentCompletionOrigin` core logic tries a bound adapter first and then a hook fallback; neither surface was wired for Slack, so completions fell back to raw `to: "channel:<channelId>"` with no `threadId`, landing in the main channel.
- Missing detection / guardrail: No integration test exercised a subagent completion delivery back into a Slack thread. Existing Slack coverage only exercised direct message-send paths.
- Contributing context (if known): The belt-and-suspenders adapter-plus-hooks pattern was already in place for Discord, Feishu, Telegram, and Matrix. Slack was simply missing from that rollout.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [ ] Unit test
  - [x] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `extensions/slack/src/thread-bindings.test.ts` (9 cases, SessionBindingAdapter exercise) and `extensions/slack/src/subagent-hooks.test.ts` (12 cases, hook-level exercise).
- Scenario the test should lock in: spawning a Slack-origin subagent stamps a thread binding, ending it unbinds, and the delivery-target hook resolves the originating thread back out (with matching, single-binding fallback, and ambiguous-binding pass-through behaviors).
- Why this is the smallest reliable guardrail: These tests run against the same `SessionBindingAdapter` and hook handlers that the production subagent delivery pipeline exercises, so they fail closed when either seam regresses. They avoid needing a live Slack workspace.
- Existing test that already covers this (if any): None for Slack. Other channels have equivalent tests (`extensions/telegram/src/thread-bindings.test.ts`, etc.).
- If no new test is added, why not: N/A — 21 new cases added.

## User-visible / Behavior Changes

- New opt-in config key `channels.slack.threadBindings.spawnSubagentSessions` (default `false`). When enabled, Slack subagents spawned from a thread will deliver completion messages back into the originating thread instead of the main channel.
- Also inherits `channels.slack.threadBindings.{enabled,idleHours,maxAgeHours,spawnAcpSessions}` from the shared thread-bindings schema, matching the existing Slack account-level / channel-level config inheritance pattern.
- Behavior is gated: with no config change, existing installs are unaffected.

## Diagram (if applicable)

```text
Before:
[user @mention in thread] -> [subagent spawns]
                          -> [subagent finishes]
                          -> [completion delivered to channel main surface] (bug)

After (with spawnSubagentSessions=true):
[user @mention in thread] -> [subagent spawns] -> [thread-binding recorded]
                          -> [subagent finishes]
                          -> [delivery_target hook resolves original thread]
                          -> [completion delivered back into the same thread]
```

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? No
- New/changed network calls? No
- Command/tool execution surface changed? No
- Data access scope changed? No

### Environment

- OS: macOS (Darwin 25.3.0)
- Runtime/container: Local gateway, Node 22+
- Model/provider: N/A (routing-only fix)
- Integration/channel (if any): Slack channel plugin, at least one Slack account
- Relevant config (redacted): `channels.slack.threadBindings.spawnSubagentSessions: true` (plus `threadBindings.enabled: true` at session or account level)

### Steps

1. Configure a Slack account with `threadBindings.spawnSubagentSessions: true`.
2. In a Slack thread, mention the bot with a prompt that spawns a subagent.
3. Wait for the subagent to finish and emit its completion message.

### Expected

- The completion message posts back into the same thread where the user mentioned the bot.

### Actual (before)

- The completion message posts into the main channel, losing thread context.

### Actual (after)

- The completion message posts back into the originating thread.

## Evidence

- [x] Failing test/log before + passing after

21 new Vitest cases (9 in `thread-bindings.test.ts`, 12 in `subagent-hooks.test.ts`) exercise the full spawn → end → delivery-target lifecycle, including ambiguous-binding pass-through and single-binding fallback. All pass against the rebased tree.

## Human Verification (required)

- Verified scenarios:
  - `pnpm test extensions/slack/src/thread-bindings.test.ts extensions/slack/src/subagent-hooks.test.ts` — 21/21 passing, both pre- and post-rebase onto latest `origin/main`.
  - `pnpm tsgo` — green, pre- and post-rebase.
  - `pnpm check` (lint, format, and all policy guards) — green.
  - `pnpm build` — green post-rebase, no `[INEFFECTIVE_DYNAMIC_IMPORT]` warnings, plugin-sdk exports verified. This is the hard gate because the change touches bundled plugin startup and lazy-loading boundaries.
  - Full `pnpm test` — one unrelated pre-existing timing race in `extensions/qa-lab/src/telegram-live.runtime.test.ts` (the test file is byte-identical to `origin/main` and my changeset does not touch qa-lab at all).
- Edge cases checked: ambiguous bindings across channels, single-binding fallback, non-Slack channel no-op, `threadRequested: false` no-op, global vs account-level `threadBindings.enabled` override, child-placement rejection (Slack cannot create empty threads without posting a parent message, so child placement returns `BINDING_CAPABILITY_UNSUPPORTED`).
- What I did NOT verify: Live end-to-end against a real Slack workspace (no live Slack credentials in this environment). The 21 seam tests exercise the same adapter and hook code paths that production uses.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? Yes
- Config/env changes? Yes (additive only — new opt-in `spawnSubagentSessions` key)
- Migration needed? No
- If yes, exact upgrade steps: N/A

## Risks and Mitigations

- Risk: When `spawnSubagentSessions` is enabled but the originating thread becomes unreachable (archived, deleted, or otherwise missing from the binding table), the delivery hook could return an empty target.
  - Mitigation: The `subagent_delivery_target` handler falls back to single-binding match when only one binding exists for the session, and keeps original routing when delivery is ambiguous across channels. Both paths have explicit test coverage in `subagent-hooks.test.ts`.
- Risk: A stale thread binding could survive past subagent completion and route a later unrelated send into the wrong thread.
  - Mitigation: `subagent_ended` unbinds by session key, and the manager supports idle-timeout and max-age sweeps inherited from the shared thread-bindings infrastructure.

### CODE REVIEW

**Comment by jaredar1232** [comment]
Pushed a follow-up commit (`3206efcf7f`) addressing the Greptile review:

**1. Duplicated `parseSlackChannelIdFromTo` (resolved)**
Exported the helper from `extensions/slack/src/thread-bindings.ts` and imported it in `subagent-hooks.ts`. The local copy is gone.

**2. `resolveThreadBindingFlags` partial duplication (partial fix)**
`resolveThreadBindingFlags` now calls `resolveThreadBindingSpawnPolicy` from `openclaw/plugin-sdk/conversation-runtime` for the `enabled` chain so the 3-tier (account → channel → session) fallback stays in sync with the core helper. `spawnSubagentSessions` is still resolved locally with an explicit `false` default — full delegation isn't safe here because the core helper defaults `spawnEnabled` to `resolveDefaultTopLevelPlacement(channel) !== "child"`, which for Slack (`defaultTopLevelPlacement: "current"`, see `extensions/slack/src/channel.ts:356`) would default it to `true`. That would silently enable subagent thread bindings on every Slack install on upgrade and break the opt-in design called out in the changelog (`with opt-in channels.slack.threadBindings.spawnSubagentSessions`). Discord follows the same local-resolution pattern for the same reason.

**3. Fallout from a new lint rule**
While running the gates after the rebase + edits, the new upstream `consistent-return` rule (`0ebeee8b0d chore: enable consistent-return`) flagged the spawning and delivery-target handlers. Same commit adds explicit return-type annotations (`SlackSubagentSpawningResult`, `SlackSubagentDeliveryTargetResult`) and converts the bare `return;` paths to `return undefined;`, mirroring the Discord handlers.

**Verification**
- `pnpm test extensions/slack/src/thread-bindings.test.ts extensions/slack/src/subagent-hooks.test.ts` — 21/21 passing
- `pnpm check` — clean
- `pnpm build` — clean, no `[INEFFECTIVE_DYNAMIC_IMPORT]` warnings

**Comment by jaredar1232** [comment]
Pushed follow-up commit `42f2c2fe81` addressing the Codex review.

**The bug was real.** `unbindSlackThreadBindingsBySessionKey` was calling `manager.unbindBySessionKey({targetSessionKey, reason})` — no kind argument — which deletes every binding for that session key from memory and persistence, then applying `.filter(entry => entry.targetKind === params.targetKind)` on the already-deleted array. In a mixed-kind session (subagent + ACP records sharing the same `targetSessionKey`), a `subagent_ended` event with `targetKind: "subagent"` would silently wipe the ACP record from state, and later delivery/routing could no longer resolve it.

**Fix.** Pushed the `targetKind` filter down into `manager.unbindBySessionKey` so the check runs during the iteration loop, before the `bindingsByAccountThread.delete(key)` call. The outer `unbindSlackThreadBindingsBySessionKey` now just passes the kind through; the post-filter is gone. The `SessionBindingAdapter.unbind` full-detach path stays unchanged — it still calls without a kind, so it still removes all bindings for the session key, which is the correct full-detach semantics.

**Test.** Added `only removes matching-kind bindings when targetKind is passed to unbindBySessionKey` — binds two records with the same `targetSessionKey` (one `subagent`, one `session`/ACP) on different threads, calls `unbindSlackThreadBindingsBySessionKey({targetKind: "subagent"})`, and asserts:
1. Return value is length 1 and matches the subagent thread.
2. `listBySessionKey` still shows the ACP entry.
3. `getByThread` still resolves the ACP thread and no longer resolves the subagent thread.

**Verification**
- `pnpm test extensions/slack/src/thread-bindings.test.ts extensions/slack/src/subagent-hooks.test.ts` — 22/22 passing (was 21; new test is the +1)
- `pnpm check` — clean
- `pnpm build` — clean, no `[INEFFECTIVE_DYNAMIC_IMPORT]` warnings

**Comment by jaredar1232** [comment]
**Re: #54840 relationship** — a maintainer flagged that #54840 (merged 2026-04-09, shipped in v2026.4.10) may have already fixed this. After testing on v2026.4.11, the symptom still reproduces. The two PRs are complementary — see [the detailed trace on the issue](https://github.com/openclaw/openclaw/issues/64454#issuecomment-4230825275).

**TL;DR:** #54840 fixed the channel gate inside `resolveConversationDeliveryTarget` so Slack thread-child sessions get `{ to: "channel:<channelId>", threadId: <threadTs> }`. That fix is correct and deployed. But `resolveConversationDeliveryTarget` is only reached when a `SessionBindingAdapter` binding exists for the child session (`route.mode === "bound"` in `resolveSubagentCompletionOrigin`). The subagent spawn path doesn't create a binding — the child's delivery context is `{"channel": "slack"}` with no `to`/`threadId` — so the bound path never activates and the thread-child branch never fires.

This PR fills the gap by registering a Slack `SessionBindingAdapter` and a `subagent_spawning` hook that writes the binding record with the correct `{channelId, threadTs}` when subagents are spawned. At completion, the binding exists, the bound delivery path activates, and #54840's generalized channel gate correctly routes the message to the originating thread.

**Comment by jaredar1232** [comment]
**Re: Codex review — spawnAcpSessions default-enabled for Slack**

Valid catch. The `conversationBindings` registration with `defaultTopLevelPlacement: "current"` causes `resolveThreadBindingSpawnPolicy` (line 207 of `src/channels/thread-bindings-policy.ts`) to default `spawnEnabled` to `true` for ACP spawns on Slack, since `"current" !== "child"` evaluates to true. Before this PR no binding adapter existed for Slack so it was moot, but now the monitor provider creates one at gateway startup.

The root issue is actually broader than the registration — even without `defaultTopLevelPlacement`, the core falls back to `"current"` for any channel (line 58), so the default would be `true` regardless. The `conversationBindings` registration doesn't change the policy outcome; it just makes the channel's intent explicit.

**Fix (3fe31d2):** Changed `spawnAcpSessions` from `.optional()` to `.default(false)` in the Slack zod schema. This ensures that when `threadBindings` is configured (as it must be for `spawnSubagentSessions: true`), ACP defaults to disabled unless explicitly opted in — matching the type documentation in `SlackThreadBindingsConfig` which says "Default: false (opt-in)".

Edge case: if `threadBindings` is entirely absent from the config, the whole object is `undefined` and the zod default doesn't apply. In that scenario the core policy still defaults ACP to `true`. However, that case is narrow — subagent bindings (the purpose of this PR) also don't work without `threadBindings`, and ACP requires an explicit `--thread` flag from the user. If full gating is needed, a core change to `resolveThreadBindingSpawnPolicy`'s default logic would be the right follow-up.

**Comment by jaredar1232** [comment]
**Re: follow-up Codex review — account-level default breaks inheritance**

Good catch. The `.default(false)` was in `SlackAccountSchema` which is shared by both root and per-account configs. Since the core policy does `account?.spawnAcpSessions ?? root?.spawnAcpSessions`, the zod-injected `false` at the account level would block inheritance from a root-level `spawnAcpSessions: true`.

**Fix (a4c5026):** Moved the `.default(false)` from `SlackAccountSchema` to the `SlackConfigSchema.safeExtend()` override. Now:
- **Root** (`channels.slack.threadBindings`): `spawnAcpSessions` defaults to `false` via `safeExtend` — so configuring `threadBindings: { spawnSubagentSessions: true }` without mentioning ACP correctly disables ACP.
- **Account** (`channels.slack.accounts.*.threadBindings`): `spawnAcpSessions` stays `.optional()` — omitting it lets the `??` chain fall through to the root value, preserving inheritance.

File: extensions/slack/src/monitor/provider.ts [hunk 1]
```diff
@@ -283,6 +289,28 @@ export async function monitorSlackProvider(opts: MonitorSlackOpts = {}) {
     );
   }
 
+  const threadBindingPolicy = resolveThreadBindingSpawnPolicy({
+    cfg,
+    channel: "slack",
+    accountId: account.accountId,
+    kind: "subagent",
```
[the remark this answers is not in the document]
Comment (jaredar1232) [reply extensions/slack/src/monitor/provider.ts:296]: This follows the existing structure of `monitorSlackProvider` — all pre-loop setup (app construction, auth test, context creation, event registration) lives before the `try`/`finally` at line 608, and the manager creation slots into that same region. If startup fails before the main loop, the process is exiting and the sweeper timer won't survive. Moving the manager inside the `try` block would require restructuring the entire function, which is out of scope for this PR. Leaving as a potential follow-up.

File: extensions/slack/src/subagent-hooks.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    binding = bindings[0];
+  }
+  if (!binding) {
+    return undefined;
+  }
+  return {
+    origin: {
+      channel: "slack" as const,
+      accountId: binding.accountId,
+      to: `channel:${binding.channelId}`,
```
[the remark this answers is not in the document]
Comment (jaredar1232) [reply extensions/slack/src/subagent-hooks.ts:225]: Valid observation, but this only affects DM subagent spawns where `channelId` stores a `user:U123` value. This PR targets channel thread routing (#64454), not DM thread delivery. The DM delivery path is the exact domain that #65286 addresses from the core side (adding `user:` to `resolveExplicitConversationTargetId` and a Slack-specific `resolveDeliveryTarget`). Fixing it here would overlap with that PR and expand scope. Leaving for a DM-focused follow-up.

File: extensions/slack/src/thread-bindings.ts [hunk 3]
```diff
@@ -0,0 +1,990 @@
...
+        bindingId,
+      });
+      if (!resolvedConversationId) {
+        return;
+      }
+      const colonIdx = resolvedConversationId.indexOf(":");
+      if (colonIdx <= 0) {
+        return;
+      }
+      const channelId = resolvedConversationId.slice(0, colonIdx).trim();
```
[the remark this answers is not in the document]
Comment (jaredar1232) [reply extensions/slack/src/thread-bindings.ts:703]: Same root cause as the DM recipient kind issue above — both stem from `channelId` containing a colon when the binding originates from a DM (`user:U123`). This only affects DM subagent spawns, not the channel thread routing this PR targets. PR #65286 addresses DM delivery from the core side; the binding-ID parsing fix should follow alongside that work so the DM path is handled end-to-end in one change.

**Comment by jaredar1232** [comment]
@martingarramon Updating like you had mentioned previously to the most recent version did not work, Claude left a reply explaining why somewhere in this thread. I think this PR is good to go.

I do Operations now, so while I do still build a lot of stuff, I mainly vibe code things, including this PR. Would appreciate eyes from someone still actively programming. 

Openclaw is integrated in our medium scale business and this bug has had real impacts and is annoying to deal with since our setup frequently uses sub agents so it fails to track it's replies in threads. Just hoping to get things working!

Sorry if you're not the person to tag, I haven't made a PR to open source repo's before so not sure the the standard practice.

**Comment by jaredar1232** [comment]
Closing in favor of #64864, which takes a simpler stateless approach that covers the fix without the full binding-manager infrastructure.
